src/JsonSchema/SchemaFactory.php

```diff
@@ -230,14 +233,20 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
                 break;
             }
 
-            if ($type->isNullable()) {
-                $propertySchema['anyOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
-            } else {
-                $propertySchema['$ref'] = $subSchema['$ref'];
-            }
+            $refs[] = ['$ref' => $subSchema['$ref']];
+            $isNullable = $isNullable ?? $type->isNullable();
+        }
 
+        if ($isNullable) {
+            $refs[] = ['type' => 'null'];
+        }
+
+        if (\count($refs) > 1) {
```

Comment:
can we keep the \count in a variable to avoid counting twice?z
```suggestion
        if (($c = \count($refs)) > 1) {
```

use `(1 === $c)` later on

Reply:
Done!